### PR TITLE
IPv6 support

### PR DESCRIPTION
**Pull request by KeRan213539**
https://www.yuque.com/docs/share/864a401b-e905-4dda-9e6f-72c0c127b811?#（密码：yrol） 《Nacos支持ipv6网络环境的改造》

### CODE REVIEW

File: naming/src/main/java/com/alibaba/nacos/naming/core/Instance.java [hunk 1]
```diff
@@ -117,19 +115,18 @@ public static Instance fromString(String config) {
         }
         
         String provider = ipAddressAttributes[0];
-        Matcher matcher = IP_PATTERN.matcher(provider);
-        if (!matcher.matches()) {
+        String[] providerAddr = IpUtil.splitIpPortStr(provider);
+        if (providerAddr.length != IpUtil.SPLIT_IP_PORT_RESULT_LENGTH) {
+            // not ip:port string
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/Instance.java:120]: Is the same behavior when provider without port like `11.11.11.11` ?

File: common/src/main/java/com/alibaba/nacos/common/utils/IpUtil.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    }
+    
+    /**
+     * Check if the address contains a port.
+     * 2020/9/3 14:53
+     * @param address address string
+     * @return boolean
+     */
+    public static boolean containsPort(String address) {
+        return splitIpPortStr(address).length == 2;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/utils/IpUtil.java:106]: Can `2` replaced with `SPLIT_IP_PORT_RESULT_LENGTH`

File: naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java [hunk 3]
```diff
@@ -796,7 +797,8 @@ public int getPagedService(String namespaceId, int startPage, int pageSize, Stri
                 contained = false;
                 List<Instance> instances = service.allIPs();
                 for (Instance instance : instances) {
-                    if (containedInstance.contains(":")) {
+                    String[] containedInstanceIpPort = IpUtil.splitIpPortStr(containedInstance);
+                    if (containedInstanceIpPort.length == IpUtil.SPLIT_IP_PORT_RESULT_LENGTH) {
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:801]: replaced with `IpUtil.containPort`?

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: naming/src/main/java/com/alibaba/nacos/naming/core/Instance.java [hunk 4]
```diff
@@ -117,19 +115,19 @@ public static Instance fromString(String config) {
...
+        } catch (Exception ex) {
             return null;
         }
         
-        int expectedGroupCount = 2;
-        
         int port = 0;
-        if (NumberUtils.isNumber(matcher.group(expectedGroupCount))) {
-            port = Integer.parseInt(matcher.group(expectedGroupCount));
+        if (NumberUtils.isNumber(providerAddr[1])) {
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/Instance.java:126]: possible for OutOfIndexException？

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Review by KomachiSion — DISMISSED** [review]

File: common/src/test/java/com/alibaba/nacos/common/utils/IpUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+/**
+ * test IpUtil.
+ * @ClassName: IpUtilTest
+ * @date 2020/9/3 10:31
+ */
+public class IpUtilTest {
+    
+    @Test
+    public void test111(){
```
Comment (KomachiSion) [inline common/src/test/java/com/alibaba/nacos/common/utils/IpUtilTest.java:34]: format test method name to describe what situation the method test

File: common/src/test/java/com/alibaba/nacos/common/utils/IpUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces();
+            while (en.hasMoreElements()) {
+                NetworkInterface ni = en.nextElement();
+                Enumeration<InetAddress> ads = ni.getInetAddresses();
+                while (ads.hasMoreElements()) {
+                    InetAddress ip = ads.nextElement();
+                    System.out.println(ip.getHostAddress());
+                    // Compatible group does not regulate 11 network segments
+                    if (!ip.isLoopbackAddress() && ip.getHostAddress().indexOf(":") == -1
+                        /* && ip.isSiteLocalAddress() */) {
```
Comment (KomachiSion) [inline common/src/test/java/com/alibaba/nacos/common/utils/IpUtilTest.java:45]: remove unused codes

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Review by KomachiSion — APPROVED** [review]

File: address/src/main/java/com/alibaba/nacos/address/controller/AddressServerClusterController.java [hunk 7]
```diff
@@ -87,8 +87,8 @@ public ResponseEntity postCluster(@RequestParam(required = false) String product
             clusterObj.setHealthChecker(new AbstractHealthChecker.None());
             serviceManager.createServiceIfAbsent(Constants.DEFAULT_NAMESPACE_ID, serviceName, false, clusterObj);
             String[] ipArray = addressServerManager.splitIps(ips);
-            String checkResult = AddressServerParamCheckUtil.checkIps(ipArray);
-            if (AddressServerParamCheckUtil.CHECK_OK.equals(checkResult)) {
+            String checkResult = IpUtil.checkIps(ipArray);
+            if (IpUtil.CHECK_OK.equals(checkResult)) {
```
Comment (chuntaojun) [inline address/src/main/java/com/alibaba/nacos/address/controller/AddressServerClusterController.java:91]: IpUtil.checkOK(str) I think is better

File: common/src/main/java/com/alibaba/nacos/common/tls/SelfHostnameVerifier.java [hunk 8]
```diff
@@ -38,7 +38,7 @@
     
     private static ConcurrentHashMap<String, Boolean> hosts = new ConcurrentHashMap<String, Boolean>();
     
-    private static final String[] LOCALHOST_HOSTNAME = new String[] {"localhost", "127.0.0.1"};
+    private static final String[] LOCALHOST_HOSTNAME = new String[] {"localhost", IpUtil.localHostIp()};
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/tls/SelfHostnameVerifier.java:41]: Ip => IP

File: common/src/main/java/com/alibaba/nacos/common/utils/IpUtil.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+import java.net.UnknownHostException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * ip tool.
+ *
+ * @author Nacos
+ */
+public class IpUtil {
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/utils/IpUtil.java:29]: IpUtil => IPUtil

File: client/src/main/java/com/alibaba/nacos/client/config/utils/ParamUtils.java [hunk 10]
```diff
@@ -190,7 +190,7 @@ public static void checkBetaIps(String betaIps) throws NacosException {
         }
         String[] ipsArr = betaIps.split(",");
         for (String ip : ipsArr) {
-            if (!IpUtils.isIpv4(ip)) {
+            if (!IpUtil.isIpv4(ip)) {
```
Comment (wangweizZZ) [inline client/src/main/java/com/alibaba/nacos/client/config/utils/ParamUtils.java:193]: 这种地方现在是不是应该用isIp替换isIpv4了？

File: core/src/main/java/com/alibaba/nacos/core/utils/WebUtils.java [hunk 11]
```diff
@@ -93,6 +93,11 @@ private static String resolveValue(String value, String encoding) {
         try {
             value = HttpUtils.decode(new String(value.getBytes(StandardCharsets.UTF_8), encoding), encoding);
         } catch (UnsupportedEncodingException ignore) {
+        } catch (Exception ex) {
```
[the remark this answers is not in the document]
Comment (KeRan213539) [reply core/src/main/java/com/alibaba/nacos/core/utils/WebUtils.java:116]: 解决重复decode的问题

File: core/src/main/java/com/alibaba/nacos/core/utils/WebUtils.java [hunk 12]
```diff
@@ -42,7 +42,7 @@
  * @author nkorange
  */
 public class WebUtils {
-    
+
```
Comment (Maijh97) [inline core/src/main/java/com/alibaba/nacos/core/utils/WebUtils.java:45]: Please do not change the code style.

File: naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java [hunk 13]
```diff
@@ -955,7 +956,7 @@ public int getPagedService(String namespaceId, int startPage, int pageSize, Stri
                 contained = false;
                 List<Instance> instances = service.allIPs();
                 for (Instance instance : instances) {
-                    if (containedInstance.contains(":")) {
+                    if (IPUtil.containsPort(containedInstance)) {
                         if (StringUtils.equals(instance.getIp() + ":" + instance.getPort(), containedInstance)) {
```
Comment (Maijh97) [inline naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:960]: Here ":" can use IPUtil.IP_PORT_SPLITER.

File: naming/src/main/java/com/alibaba/nacos/naming/healthcheck/ClientBeatCheckTask.java [hunk 14]
```diff
@@ -135,7 +136,7 @@ private void deleteIp(Instance instance) {
                     .appendParam("ephemeral", "true").appendParam("clusterName", instance.getClusterName())
                     .appendParam("serviceName", service.getName()).appendParam("namespaceId", service.getNamespaceId());
             
-            String url = "http://127.0.0.1:" + ApplicationUtils.getPort() + ApplicationUtils.getContextPath()
+            String url = "http://" + IPUtil.localHostIP() + ":" + ApplicationUtils.getPort() + ApplicationUtils.getContextPath()
```
Comment (Maijh97) [inline naming/src/main/java/com/alibaba/nacos/naming/healthcheck/ClientBeatCheckTask.java:139]: Here ":" can use IPUtil.IP_PORT_SPLITER.

File: sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java [hunk 15]
```diff
@@ -87,20 +81,20 @@ public void run() {
...
+                if (!StringUtils.isBlank(nacosIP)) {
+                    if (!(IPUtil.isIP(nacosIP) || isDomain(nacosIP))) {
+                        throw new RuntimeException("nacos address " + nacosIP + " is not ip");
+                    }
                 }
                 String tmpSelfIP = nacosIP;
                 if (StringUtils.isBlank(tmpSelfIP)) {
                     preferHostnameOverIP = Boolean.getBoolean(SYSTEM_PREFER_HOSTNAME_OVER_IP);
-                    
+
```
Comment (Maijh97) [inline sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java:92]: Please do not change the code style.

File: sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java [hunk 16]
```diff
@@ -87,20 +81,20 @@ public void run() {
...
                 if (StringUtils.isBlank(tmpSelfIP)) {
                     preferHostnameOverIP = Boolean.getBoolean(SYSTEM_PREFER_HOSTNAME_OVER_IP);
-                    
+    
                     if (!preferHostnameOverIP) {
                         preferHostnameOverIP = Boolean
                                 .parseBoolean(ApplicationUtils.getProperty(PREFER_HOSTNAME_OVER_IP));
                     }
-                    
+
```
Comment (Maijh97) [inline sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java:97]: Please do not change the code style.

File: sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java [hunk 17]
```diff
@@ -159,8 +160,9 @@ public static InetAddress findFirstNonLoopbackAddress() {
                     if (!ignoreInterface(ifc.getDisplayName())) {
                         for (Enumeration<InetAddress> addrs = ifc.getInetAddresses(); addrs.hasMoreElements(); ) {
                             InetAddress address = addrs.nextElement();
-                            if (address instanceof Inet4Address && !address.isLoopbackAddress() && isPreferredAddress(
-                                    address)) {
+                            if ((IPUtil.PREFER_IPV6_ADDRESSES ? address instanceof Inet6Address
+                                    : address instanceof Inet4Address) && !address.isLoopbackAddress()
+                                    && isPreferredAddress(address)) {
```
Comment (Maijh97) [inline sys/src/main/java/com/alibaba/nacos/sys/utils/InetUtils.java:165]: `IPUtil.PREFER_IPV6_ADDRESSES ? address instanceof Inet6Address : address instanceof Inet4Address`  
Here is a separate line, which is more readable.

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**